csharp/extractor/Semmle.Extraction.CSharp/Entities/Types/Type.cs

```diff
@@ -347,6 +355,211 @@ public override bool Equals(object? obj)
+                            workQueue.Enqueue(namedTypeArgument.OriginalDefinition);
+                        }
+                    }
+                }
+            }
+
+            private bool HasExpandingCycle()
+            {
+                return startSymbol.TypeParameters.Any(HasExpandingCycle);
+            }
+
+            private bool HasExpandingCycle(ITypeParameterSymbol start)
+            {
+                var visited = new HashSet<ITypeParameterSymbol>(SymbolEqualityComparer.Default);
+                var recStack = new HashSet<ITypeParameterSymbol>(SymbolEqualityComparer.Default);
+                var hasExpandingCycle = HasExpandingCycle(start, visited, recStack, start, hasSeenExpandingEdge: false);
+                return hasExpandingCycle;
+            }
+
+            private List<(ITypeParameterSymbol To, bool IsExpanding)> GetOutgoingEdges(ITypeParameterSymbol typeParameter)
+            {
+                return edges.TryGetValue(typeParameter, out var outgoingEdges)
+                    ? outgoingEdges
+                    : new List<(ITypeParameterSymbol, bool)>();
+            }
+
+            /// <summary>
+            /// A modified cycle detection algorithm based on DFS.
+            /// </summary>
+            /// <param name="current">The current node that is being visited</param>
+            /// <param name="visited">The nodes that have already been visited by any path.</param>
+            /// <param name="currentPath">The nodes already visited on the current path. Could be a List<> if the order was important.</param>
+            /// <param name="start">The start and end of the cycle. We're not looking for any cycle, but a cycle that goes back to the start.</param>
+            /// <param name="hasSeenExpandingEdge">Whether an expanding edge was already seen in this path. We're looking for a cycle that has at least one expanding edge.</param>
+            /// <returns></returns>
+            private bool HasExpandingCycle(ITypeParameterSymbol current, HashSet<ITypeParameterSymbol> visited, HashSet<ITypeParameterSymbol> currentPath, ITypeParameterSymbol start, bool hasSeenExpandingEdge)
+            {
+                if (currentPath.Count > 0 && SymbolEqualityComparer.Default.Equals(current, start))
+                {
+                    return hasSeenExpandingEdge;
+                }
+
+                if (visited.Contains(current))
+                {
+                    return false;
+                }
+
+                visited.Add(current);
+                currentPath.Add(current);
+
+                var outgoingEdges = GetOutgoingEdges(current);
+
+                foreach (var outgoingEdge in outgoingEdges)
+                {
+                    if (HasExpandingCycle(outgoingEdge.To, visited, currentPath, start, hasSeenExpandingEdge: hasSeenExpandingEdge || outgoingEdge.IsExpanding))
+                    {
+                        return true;
+                    }
+                }
```

Comment:
```suggestion
                if (GetOutgoingEdges(current).Any(outgoingEdge => HasExpandingCycle(outgoingEdge.To, visited, currentPath, start, hasSeenExpandingEdge: hasSeenExpandingEdge || outgoingEdge.IsExpanding)))
                {
                    return true;
                }

Reply:
I'll keep this as it is, I think it's easier to read the `foreach-if` than the `if-Any`. Also, the former is probably slightly more performant.